### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by gudrutis**
This pull request will update script cmsBuild in the following ways:
* It will allow overwriting source location of the package to a local directory;
* It will output formatted links for sources ( may be useful for developers).

@smuzaffar , I suggest having a development discussion to this PR and only merge when we are finished developing.

### CODE REVIEW

**Comment by smuzaffar** [comment]
lets get this merged for V00-32-XX-devel branch and run some tests

**Comment by smuzaffar** [comment]
@gudrutis , looks like you have missed some changes, where are changes to check for local sources?

File: cmsBuild [hunk 1]
```diff
@@ -442,78 +471,82 @@ def downloadTc (source, dest, options):
...
+                else:
+                    pip_opts = pip_opts + ' ' + spSrc[i]
+            break
+    comm = ""
+    if pip_package_env_script: comm = ". %s; " % pip_package_env_script
+    comm = comm + ' which pip; pip download ' + pip_opts + ' -d ' + tempdir + ' ' + pack + '; /bin/ls ' + tempdir + '| grep -v files.list > ' + tempdir + '/files.list'
+    error, output = getstatusoutput(comm)
+    if error:
+        log("Error while downloading sources from %s using pip.\n\n %s \n %s \n  %s \n " % (
+        source, comm, error, output))
```
Comment (smuzaffar) [inline cmsBuild:538]: @gudrutis , shouldn't this line have some extra indentation?

Comment (gudrutis) [reply cmsBuild:538]: @smuzaffar True, it looks like it flew under some IDEA radar. I will recheck in a few commits.
 
I also noticed there are a lot of unused imports, should I clean up?

**Comment by gudrutis** [comment]
> @gudrutis , looks like you have missed some changes, where are changes to check for local sources?

That part of the function is still under development, so I did not commit. Have a few questions, so I thought I would ask when you here when I would finish a bit of structuring.

File: cmsBuild (hunk 1 above)
Comment (smuzaffar) [reply cmsBuild:538]: yes please.

**Comment by gudrutis** [comment]
Ah, since the pull request is closed, latter commits do not show up here

File: cmsBuild [hunk 2]
```diff
@@ -2520,6 +2535,14 @@ def parseOptions():
                       help="When uploading, copy the temporary directory to the production one.",
                       default=False)
     advancedBuildOptions = OptionGroup(parser, "Advanced build options")
+    advancedBuildOptions.add_option("--source",
+                                    dest="localSources",
+                                    action="append",
+                                    # nargs='+', works only with newer argpars
+                                    # type="str",
```
Comment (smuzaffar) [inline cmsBuild]: remove the commented out code

File: cmsBuild [hunk 3]
```diff
@@ -3782,6 +3805,42 @@ def rpmCacheUpdate(opts, cache=None, rpm_env=None):
...
+        pkg = source = None
+        if ":" in pkg_and_source:
+            pkg, source = pkg_and_source.split(":")
+        else:
+            pkg = pkg_and_source
+            source = "Source"
+        #check if given directory path is correct
+        if not exists(dir_path):
+            fatal("WARNING: the following direcotry for package %s and source %s does not exist: %s" % (pkg, source, dir_path))
+        if not os.path.isdir(dir_path):
```
Comment (smuzaffar) [inline cmsBuild:3836]: as discussed, we should allow to pass a file path too. so check about directory and empty directory can be relaxed here

File: cmsBuild [hunk 4]
```diff
@@ -2535,11 +2535,15 @@ def parseOptions():
                       help="When uploading, copy the temporary directory to the production one.",
                       default=False)
     advancedBuildOptions = OptionGroup(parser, "Advanced build options")
+    # TODO-zm for dumping out sources
```
Comment (smuzaffar) [inline cmsBuild:2538]: do we need this comment any more?

**Comment by smuzaffar** [comment]
I wanted to test it , so I have merged it in to V00-32-XX-devel branch :) Open a new PR for the devel branch

File: cmsBuild (hunk 4 above)
Comment (gudrutis) [reply cmsBuild:2538]: Ok, will be deleted.

File: cmsBuild (hunk 2 above)
Comment (gudrutis) [reply cmsBuild]: I think you were checking wrong commit, in this one comments are already removed
https://github.com/cms-sw/pkgtools/pull/141/commits/614bfeee5033056aa041a03e90591ed3b22d07c6

File: cmsBuild (hunk 3 above)
Comment (gudrutis) [reply cmsBuild:3836]: https://github.com/gudrutis/pkgtools/commit/0ec3662095d9de63c93c99e9e36d187f2223c3c8

In this commit I allow files and directories, but if it is an empty directory, I will fail it.

**Pull request merged**